### ISSUE TITLE
Question: full restore of a cluster (DR scenario)

### ISSUE THREAD

**Issue by amolari** [F5Networks/f5-aws-cloudformation#121]
## Do you already have an issue opened with F5 support?
No

## Description
Refering to the documentation: https://github.com/F5Networks/f5-aws-cloudformation/tree/master/supported/autoscale/ltm/via-lb/1nic/existing-stack/bigiq#restoring-or-upgrading-the-solution
If we have a backup of the cluster in a safe place, what could be the process to fully restore the configuration on a new cluster (DR-scenario)? If we upload that UCS on the new Bucket as soon as that bucket and its /backup directory have been created (stack building), would the Primary instance use that backup? Are there any limitation to take into account?

## Template
f5-aws-cloudformation/supported/autoscale/ltm/via-lb/1nic/existing-stack/bigiq/
v5.7.0

## Severity Level
Severity: 5

**Comment by andreykashcheev** [comment]
Here are steps to restore cluster configuration after updating Autoscale Stack. As example, it describes steps needed to update BIGIP version (aka AMI ids) 

  1) Prepare backup (aka UCS) file:
       - connect to primary host
       - execute "crontab -e" to find cronjob used for triggering backup process (UCS generation). The back script is `/config/cloud/aws/run_autoscale_backup.sh`
       - execute the bash script used for generating UCS file. The script handles generation and upload of UCS file to cloud storage (S3 bucket) used by the autoscale stack.
 2) Update Template with new AMI Ids
 3) Update Cloud Formation stack with new template; the following AWS doc provides details on how to update existent stack:
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/updating.stacks.walkthrough.html
  4) At this point, Cloud Formation stack should have updated Template with AMI ids, so all new VMs would be used new BIGIP version. Now, it is time to spin out VMs using new AMI Ids
      - Update AutoscaleGroup policy to scale-in to 0 hosts. NOTE: the primary host includes scale-in protection, which needs to be removed in order to allow host termination
      - Update AutoscaleGroup policy to scale-out to 2 (or any other desire number of vms) 
      - New VMs would be provisioned using updated template with new AMI ids. The new primary host would elected based on the lowest mgmt private ip and the primary host would restore UCS with cluster config as part of becomePrimary workflow. 

Let me know if you have any questions/concerns

**Comment by amolari** [comment]
@andreykashcheev My question was rather on the DR scenario: we have a backup UCS of a working configuration. For wathever reason, we need to destroy the existing cluster and would like to create a new cluster (so new CF stack) and have that UCS restored during the provisioning.

**Comment by andreykashcheev** [comment]
Sorry, I mis-read/mis-understood the question.

For the DR case (when there is an entirely new stack), you will need to use the following steps: 
  1) As a part of provisioning Autoscale Solution, s3 bucket will be created.
  2) When bucket would be created you will need to upload UCS file under /backup folder 
  3) If UCS file would be uploaded before primary host is elected, then the new primary host would download and restore UCS file as part of becomePrimary workflow:
https://github.com/F5Networks/f5-cloud-libs/blob/master/scripts/autoscale.js#L1139
  4) In a case, when primary host gets elected before UCS file is uploaded under /backup folder, you can trigger re-election by deleting all existent instance metadata stored under /instance folder. Since there would be no instance metadata, autoscale cluster will re-initiate primary host election. 

Let me know if you have any questions/concerns

**Comment by amolari** [comment]
@andreykashcheev The DR scenario doesn't work because of the master key, which is different on the new instance than the one from the ucs.
I've created case 1-6585628433

Question: the UCS must be with the same naming convention (`ucsAutosave_<timestamp>.ucs`) to be found?

**Comment by andreykashcheev** [comment]
no - the ucs filename must only ends with ".ucs" in order to be retrieved and the script takes the newest UCS based on Last-Modified metadata:
https://github.com/F5Networks/f5-cloud-libs-aws/blob/master/lib/awsCloudProvider.js#L989

Before loading UCS, the autoscale script will write down UCS data into the file:
https://github.com/F5Networks/f5-cloud-libs/blob/master/scripts/autoscale.js#L1707
UCS file will be stored locally using `ucsAutosave_<timestamp>.ucs` name and it will be stored under the following directory:
https://github.com/F5Networks/f5-cloud-libs/blob/master/lib/sharedConstants.js#L51

After taking a look at ticket number 1-6585628433, I suspect that you provided incorrect ticket number.

**Comment by amolari** [comment]
@andreykashcheev case number is now correct (1-6585139310)

**Comment by andreykashcheev** [comment]
After looking at provided log snippet, it appears that the problem happens with reading extracted files from UCS file. 

**Details around the workflow:**
1) After UCS file was downloaded from S3 bucket, the solution attempts to load UCS file. In a case when UCS load fails, the solution attempts a workaround using restorePrimarykey
https://github.com/F5Networks/f5-cloud-libs/blob/59f659cb92189bb6337be64256781cbcda3db2f6/lib/bigIp.js#L1532

It appears that the UCS load failed. 

2) When solution attempted to use workaround with restoring primary key, It appears that no content was extracted from UCS because otherwise files would be read by script:
https://github.com/F5Networks/f5-cloud-libs/blob/59f659cb92189bb6337be64256781cbcda3db2f6/lib/bigIp.js#L1462

I would suggest to confirm that UCS file is valid and includes content using the following command:

```tar -tvf <ucs_filename>```

Also, while testing DR case, i would suggest to do the following steps:
  1) Wait until BIGIP would be onboarded and running
  2) Confirm that UCS is valid and it includes content
  3) Upload UCS to S3 bucket under /backup directory
  4) Delete instance metadata (all documents under S3 bucket /instances directory) to trigger re-clustering 

Let me know if you have any questions/concerns

**Comment by andreykashcheev** [comment]
@amolari 

I was able to replicate the error: 
```
2020-08-14T03:22:18.427Z silly: [pid: 24313] [lib/bigIp.js] restorePrimaryKey: bigstart stopped 
2020-08-14T03:22:18.428Z silly: [pid: 24313] [lib/bigIp.js] restorePrimaryKey: mkdir succeeded 
2020-08-14T03:22:18.429Z silly: [pid: 24313] [lib/bigIp.js] restorePrimaryKey: untarring ucs /shared/tmp/ucs/ucsUpdated_1597375174194.ucs 
2020-08-14T03:22:18.908Z silly: [pid: 24313] [lib/bigIp.js] restorePrimaryKey: untar success, reading key 
2020-08-14T03:22:18.909Z debug: [pid: 24313] [lib/bigIp.js] error restoring primary key ENOENT: no such file or directory, open '/config/tempUcs/config/bigip/kstore/primary' 
2020-08-14T03:22:18.911Z info: [pid: 24313] [scripts/autoscale.js] error loading ucs {"message":"ENOENT: no such file or directory, open '/config/tempUcs/config/bigip/kstore/primary'","stack":"Error: ENOENT: no such file or directory, open '/config/tempUcs/config/bigip/kstore/primary'","errno":-2,"code":"ENOENT","syscall":"open","path":"/config/tempUcs/config/bigip/kstore/primary"}
2020-08-14T03:22:18.913Z info: [pid: 24313] [lib/util.js] Autoscale finished. 
2020-08-14T03:22:18.910Z info: [pid: 24313] [lib/bigIp.js] loadUcs failed {"message":"ENOENT: no such file or directory, open '/config/tempUcs/config/bigip/kstore/primary'","stack":"Error: ENOENT: no such file or directory, open '/config/tempUcs/config/bigip/kstore/primary'","errno":-2,"code":"ENOENT","syscall":"open","path":"/config/tempUcs/config/bigip/kstore/primary"}
2020-08-14T03:22:18.911Z error: [pid: 24313] [scripts/autoscale.js] autoscaling error code: ENOENT message: ENOENT: no such file or directory, open '/config/tempUcs/config/bigip/kstore/primary' 
```

This caused by recent change intended to rename "master" to "primary". The problem is that BGIP still have named as "/config/tempUcs/config/bigip/kstore/master" whereas in cloud-libs, it is looking for "/config/bigip/kstore/primary"
https://github.com/F5Networks/f5-cloud-libs/blob/master/lib/bigIp.js#L1466


I have created internal issue to fix this: ESECLDTPLT-2179

The fix would be included with the next release

Let me know if you have any questions/concerns

**Comment by amolari** [comment]
@andreykashcheev 
I've performed the steps manually to repro the future changes:
- uploaded the ucs to the bucket , folder /backup
- as soon as system was up and libraries installed, I've changed the lib file with: `sed -i 's/config\/bigip\/kstore\/primary/config\\/bigip\/kstore\/master/g' /config/cloud/aws/node_modules/@f5devcentral/f5-cloud-libs/lib/bigIp.js`
I see the process of restoring is going now further than before, but I get then the errors in autoscale.log:
```
2020-08-17T15:09:03.475Z silly: [pid: 17662] [lib/awsCloudProvider.js] isPrimary: true
2020-08-17T15:09:03.476Z debug: [pid: 17662] [lib/awsCloudProvider.js] Deleting non-primaries that are not in AWS
2020-08-17T15:09:03.478Z silly: [pid: 17662] [scripts/autoscale.js] Instance status: OK
2020-08-17T15:09:03.487Z info: [pid: 17662] [scripts/autoscale.js] Initializing BIG-IP.
2020-08-17T15:09:03.506Z info: [pid: 17662] [lib/bigIp.js] This is a BIG-IP
2020-08-17T15:09:03.509Z silly: [pid: 17662] [lib/util.js] Performing local ready check
2020-08-17T15:09:04.212Z silly: [pid: 17662] [lib/util.js] Performing local ready check
2020-08-17T15:09:04.894Z info: [pid: 17662] [lib/bigIp.js] Device initialization failed name must be a string
2020-08-17T15:09:04.897Z error: [pid: 17662] [scripts/autoscale.js] autoscaling error code: ERR_ASSERTION message: name must be a string
2020-08-17T15:09:04.898Z info: [pid: 17662] [lib/util.js] Autoscale finished.
2020-08-17T15:09:04.894Z info: [pid: 17662] [lib/localCryptoUtil.js] Error decrypting data name must be a string
2020-08-17T15:09:04.894Z info: [pid: 17662] [lib/authn.js] Unable to initialize device name must be a string

```
After the repro, I've uploaded a qkview in the above mentioned case.

**Comment by andreykashcheev** [comment]
Yep - while working on validating UCS restore, I was able to replicate this error as well. I have made the change to address this and currently, I am in process of testing the change. So far, I did about several UCS restores on new cluster and they worked fine.

**Comment by jmcalalang** [comment]
@andreykashcheev 

`Yep - while working on validating UCS restore, I was able to replicate this error as well. I have made the change to address this and currently, I am in process of testing the change.`

Is the change you made going to be retro active? something we can modify or will we need to deploy a full-stack again?

**Comment by andreykashcheev** [comment]
this change requires f5-cloud-libs release and therefore, updating Cloud Template with new f5-cloud-libs version.  In other words, you will need to update the stack with new/updated template and provision new BIGIP devices to pick up the change.

To my knowledge, we are anticipating to make out-of-band/hotfix f5-cloud-libs and Cloud Templates release this sprint. I am still in process confirming that. I will provide an update as soon as it would be available. 

Let me know if you have any questions/concerns

**Comment by amolari** [comment]
@andreykashcheev Could you share the change you made to have it work please (if simple)? I'm willing to test it too.

**Comment by andreykashcheev** [comment]
I have made two changes: 
  1) Renamed from "primary" back to "master" (it should be named as /config/bigip/kstore/master since this is BIGIP system file whereas in cloud-libs, we updated to primary)
https://github.com/F5Networks/f5-cloud-libs/blob/master/lib/bigIp.js#L1466
https://github.com/F5Networks/f5-cloud-libs/blob/master/lib/bigIp.js#L1470

  2) Update checkTask method to make promise rejection for restart case:
https://github.com/F5Networks/f5-cloud-libs/blob/master/lib/bigIp.js#L1420
To trigger the following workaround when restarts happens: https://cdn.f5.com/product/bugtracker/ID639619.html

Also, I just spoke with our PM and we agreed to make out-of-band release this sprint, so all these changes would be available by eod this week. 

Let me know if you have any questions/concerns

**Comment by andreykashcheev** [comment]
We just have published new templates v5.7.1. which includes the change intended to address this issue. 

Let me know if you have any questions/concerns

**Comment by andreykashcheev** [comment]
Closing this issue...
